### PR TITLE
Support for /config folder

### PR DESCRIPTION
**Pull request by madnificent**
The config folder is available under the `/app` folder.  When files are present in this folder, compilation happens at first boot.  During development, live reload is offered for the `/config` folder.

Current *implementation* allows for users to override config files.  This allows the app to provide a default configuration for files users may override.  I consider this experimental and would certainly like input on how desired this approach is.

### CODE REVIEW

**Comment by benjay10** [comment]
I was asked to leave my findings on this after testing its functionality and some concerns.

In a test set-up with file and folder structure like this:

```
├── app
│   ├── app.js
│   ├── config.js
│   ├── Dockerfile
│   ├── otherconfig
│   │   └── index.js
│   └── package.json
├── config
│   ├── config.js
│   ├── foo.js
│   └── index.js
└── docker-compose.yml
```

I made a simple service that would import all config files:

```
import * as config1 from "./config.js";
import * as config2 from "./config/foo.js";
import * as config3 from "./config/index.js";
import * as config4 from "./config/config.js";
import * as config5 from "./config";
import * as config6 from "./otherconfig";
```

Each file contains a string identifying the location of the imported file and the service prints out the string. The `docker-compose.yml` file contains at least a mount for the `./config` folder to `/config` and, for development mode, a mount for `./app` to `/app`.

The service exhibits different behaviour between the two modi.

**In development**

In development mode, the JavaScript template is made to watch file and folders for changes and automatically rebuild and reload. Upon starting the service, the `/config` folder contents are correctly copied in folder `app/config` in the location of the built source code, so that these files can be imported like a normal JavaScript module. Changes to the `/app/config.js` and `/config/config.js` (also the others) where picked up correctly while testing, and the source code is rebuilt and the service restarted automatically.

**In production**

In order to start the service in production mode, a new docker image needs to be built and run. In production mode, the template is not meant to watch files and folders for changes, but with the added mount-config feature, the service should pick up on changes to the separate config folder on the next startup of the service, which it does as intended.

**Potential shadowing** - response to @nvdk

When importing using ES6's module system, importing from `./config` means that either a file `./config.js` needs to be present or a folder `./config` with an `index.js` file in it. The file always takes precedence over the folder if it exists. This means that a potential file `./config/index.js` is no longer imported when a new file `./config.js` is made when importing simply `./config`. The folder is being shadowed by the new file. In the test set-up, `config5` imports the `./config.js` file, because it exists. Remove that file and `config5` will import from `./config/index.js` instead! (Using the `.cjs` extension makes no difference for this.) Importing the file like in `config3` works correctly because the full path is specified. In general, if you always import files by their most specific name, you won't see a potential name shadowing anymore.


**TL;DR**

mount-config feature works as intended during my testing. There is a potential name shadowing: importing `./config` refers to file `./config/index.js` UNLESS file `./config.js` exists, which takes precedence. Importing files with their most specific name does not lead to a potential shadowing.

**Comment by madnificent** [comment]
Thanks for the validation @benjay10.

I conclude from this that we can go ahead with this feature (it doesn't break exiting configurations) but we should validate whether this is the way in which we'd want the config folder to work, or whether we'd want another construction.

Open questions from my end, which don't need to postpone this feature as an experimental feature, include:

- Do we want to allow dependencies to be installed through the configuration and how would that work?
- Is a configuration under the app folder the way we want this to behave?
- Can we implement a similar feature in the other templates? (if this matures, it needs to become an issue on mu-semtech/project.

Thoughts are welcome.

**Comment by benjay10** [comment]
I forgot to mention another topic. Mounting a `/config` folder via Docker also shadows an already existing `/app/config` folder. This could be a useful feature: you could define default config files that you can override when necessary by mounting another folder instead of the default folder. I think you could even do this at the level of individual files.

I'm just a little bit concerned about how this could become a confusing mess.

**Review by erikap — APPROVED** [review]
